packages/engine/src/scene/components/SourceComponent.ts

```diff
@@ -49,15 +51,23 @@ export const SourceComponent = defineComponent({
     }
   },
 
-  onRemove: (entity, component) => {
-    const src = component.value
+  reactor: () => {
+    const entity = useEntityContext()
+    const sourceComponent = useComponent(entity, SourceComponent)
 
-    const entities = SourceComponent.entitiesBySource[src].filter((currentEntity) => currentEntity !== entity)
-    if (entities.length === 0) {
-      SourceComponent.entitiesBySourceState[src].set(none)
-    } else {
-      SourceComponent.entitiesBySourceState[src].set(entities)
-    }
+    useLayoutEffect(() => {
+      const source = sourceComponent.value
```

Comment:
if the source changes after the component has been created, this will no longer have the same value. we should just move this line inside the returned function

Reply:
Oh if the source can change this is setup incorrectly altogether and won't clean up properly, good catch